_source/_posts/2022-05-02-angular-microfrontend-auth.md

```diff
@@ -0,0 +1,443 @@
+
+```ts
+const routes: Routes = [
+  { path: '', component: ProductsComponent },
+  { path: 'basket', loadChildren: () => import('mfeBasket/Module').then(m => m.BasketModule) },
+  { path: 'profile', loadChildren: () => import('mfeProfile/Module').then(m => m.ProfileModule)},
+  { path: 'login/callback', component: OktaCallbackComponent }
+];
+```
+
+What's this!? The IDE flags the import path as an error! The `shell` application code doesn't know about the Profile module, and TypeScript needs a little help. Open `projects/shell/src/decl.d.ts` and add the following line of code.
+
+```ts
+declare module 'mfeProfile/Module';
+```
+
+The IDE should be happier now. 😀
+
+Next, update the navigation button for **Profile** in the `shell` application to route to the correct path. Open `projects/shell/src/app/app.component.html` and find the `routerLink` for the **Profile** button. It should be approximately on line 38. Right now the `routerLink` configuration is `routerLink="/"`, but it should now be
+
+```html
+<a routerLink="/profile">
+```
+
+This is everything we need to do to connect the micro-frontend remote to the host application, but we also want to share authenticated state. Module Federation makes sharing state a piece of (cup) cake.
+
+## Shared authenticated state within micro-frontend applications
+
+To share a library, you need to configure the library in the `webpack.config.js` to share. Let's start with `shell`. Open `projects/shell/src/webpack.config.js`.
+
+There are two places to add shared code. One place is for code implementation within the project, and one is for shared external libraries. In this case, we can share the Okta external libraries as we didn't implement a service that wraps Okta's auth libraries, but I will point out both places.
+
+First, we'll add the Okta libraries. Scroll down towards the bottom of the file to the `shared` property. You'll follow the same pattern as the `@angular` libraries already in the list and add the singleton instances of the two Okta libraries as shown in this snippet:
+
+```js
+shared: share({
+  // other Angular libraries remain in the config. This is just a snippet
+  "@angular/router": { singleton: true, strictVersion: true, requiredVersion: 'auto' },
+  "@okta/okta-angular": { singleton: true, strictVersion: true, requiredVersion: 'auto' },
+  "@okta/okta-auth-js": { singleton: true, strictVersion: true, requiredVersion: 'auto' },
+
+  ...sharedMappings.getDescriptors()
+})
+```
+
+If you created a library within this project, like the basket service and project service in the starter code, you'd add the library to the `sharedMappings` array at the top of the `webpack.config.js` file. If you create a new library to wrap Okta's libraries, this is where you'd add it.
```

Comment:
```suggestion
If you created a library within this project, like the basket service and project service in the starter code, you'd add the library to the `sharedMappings` array at the top of the `webpack.config.js` file. If you created a new library to wrap Okta's libraries, this is where you'd add it.
```
seems like both "If you created.." should have the same verb tense, but I might be wrong

Reply:
@havi how does this sound (since we didn't create a wrapper I'm trying to draw the distinction)
> When you create a library within this project, like the basket service and project service in the starter code, you'd add the library to the `sharedMappings` array at the top of the `webpack.config.js` file. If you create a new library to wrap Okta's libraries, this is where you'd add it.

Reply:
wfm!